docs/Contributing.asciidoc

```diff
@@ -167,14 +167,21 @@ variable specified as a white-space separated list, for example:
 
 [source,sh]
 ----
-make test TEST=t/config.t
+make test TESTS=t/config.t
 ----
 
 or
 
 [source,sh]
 ----
-make test TEST="t/foo.t t/bar.t"
+make test TESTS="t/foo.t t/bar.t"
+----
+
+To run only unit test itself without other tests (perltidy or PostgreSQL tests):
+
+[source,sh]
+----
+make test-unit-and-integration TESTS=t/foo.t
```

Comment:
I don't see the difference to the above. `make test TESTS=t/my-test.t` should not call perltidy or "PostgreSQL tests". Also:

1. it should say "To run only a unit test" or "To run only unit tests"
2. We don't use the term "PostgreSQL tests" anywhere else and actually I don't even know what you mean with that.

Reply:
Just try `t/40-script_openqa-clone-custom-git-refspec.t`:

```
$ make test TESTS=t/40-script_openqa-clone-custom-git-refspec.t
test -d /dev/shm/tpg && (pg_ctl -D /dev/shm/tpg -s status >&/dev/null || pg_ctl -D /dev/shm/tpg -s start) || ./t/test_postgresql /dev/shm/tpg
./t/test_postgresql: line 15: initdb: command not found
make: *** [Makefile:163: test-with-database] Error 127

```
vs.

```
$ make test-unit-and-integration TESTS=t/40-script_openqa-clone-custom-git-refspec.t
export GLOBIGNORE="";\
timeout -v $((60 * (0 + 1) ))m script/retry prove -l -v t/40-script_openqa-clone-custom-git-refspec.t
t/40-script_openqa-clone-custom-git-refspec.t .. 
# Calling ' unshare -r -n script/openqa-clone-custom-git-refspec '
ok 1 - hint shown for mandatory parameter missing
ok 2 - openqa-clone-custom-git-refspec needs parameters
# Calling ' unshare -r -n script/openqa-clone-custom-git-refspec --help'
ok 3 - help text shown
ok 4 - help screen is regarded as success
# Calling ' unshare -r -n script/openqa-clone-custom-git-refspec --invalid-arg'
ok 5 - invalid args also yield help
ok 6 - help screen still success
# Calling ' unshare -r -n script/openqa-clone-custom-git-refspec https://github.com/user/repo/pull/9128 https://openqa.opensuse.org/tests/1234'
ok 7 - without network we fail (without error)
# Calling ' unshare -r -n script/openqa-clone-custom-git-refspec https://github.com/user/repo/pull/9128 https://openqa.opensuse.org/tests/1234'
ok 8 - clone-job command line is correct
ok 9 - command exits successfully
# Calling ' unshare -r -n script/openqa-clone-custom-git-refspec -v https://github.com/user/repo/pull/9128 https://openqa.opensuse.org/tests/1234'
ok 10 - clone-job with -v prints commands
ok 11 - command exits successfully
# Calling ' unshare -r -n script/openqa-clone-custom-git-refspec -n -v https://github.com/user/repo/pull/9128 https://openqa.opensuse.org/tests/1234'
ok 12 - clone-job with -n -v prints commands
ok 13 - command exits successfully
# Calling ' unshare -r -n script/openqa-clone-custom-git-refspec https://github.com/user/repo/tree/my_branch https://openqa.opensuse.org/tests/1234 FOO=bar'
ok 14 - alternative mode with branch reference also yields right variables
ok 15 - command exits successfully
# Calling 'env repo_name=user/repo pr=9128 host=https://openqa.opensuse.org job=1234  unshare -r -n script/openqa-clone-custom-git-refspec '
ok 16 - environment variables can be used instead
ok 17 - exits successfully
# Calling 'env repo_name=user/repo pr=9128 host=https://openqa.opensuse.org job=1234 testsuite=new_test needles_dir=/my/needles productdir=my/product  unshare -r -n script/openqa-clone-custom-git-refspec '
ok 18 - testsuite and dirs can be overridden
ok 19 - exits successfully
# Calling ' unshare -r -n script/openqa-clone-custom-git-refspec https://github.com/me/repo/pull/1/ https://openqa.opensuse.org/tests/1'
ok 20 - trailing slash ignored
ok 21 - command exits successfully
# Calling ' unshare -r -n script/openqa-clone-custom-git-refspec https://github.com/user/repo/pull/9128 https://openqa.opensuse.org/tests/1234,https://openqa.opensuse.org/tests/1234'
ok 22 - accepts comma-separated list of jobs
ok 23 - command exits successfully
# Calling ' unshare -r -n script/openqa-clone-custom-git-refspec https://github.com/user/repo/pull/9128 https://openqa.opensuse.org/tests/1234,https://openqa.opensuse.org/tests/1234 FOO=bar'
ok 24 - additional arguments are passed as test parameters for each job
ok 25 - command exits successfully
# Calling ' unshare -r -n script/openqa-clone-custom-git-refspec --clone-job-args="--show-progress" https://github.com/user/repo/pull/9128 https://openqa.opensuse.org/tests/1234'
ok 26 - additional parameters can be passed to clone-job
ok 27 - command exits successfully
# Calling 'dry_run='printf "%s\n"'  unshare -r -n script/openqa-clone-custom-git-refspec https://github.com/me/repo/pull/1/ https://openqa.opensuse.org/tests/1 TEST1=$FOO_BAR 'TEST2=$VAR' TEST3=space\ space 'TEST4=(!?bla)''
ok 28 - Custom variables has proper bash escaping
# Calling ' unshare -r -n script/openqa-clone-custom-git-refspec https://github.com/user/repo/pull/9128 https://openqa.opensuse.org/t1234'
not ok 29 - short test URLs are supported the same # TODO not implemented
#   Failed (TODO) test 'short test URLs are supported the same'
#   at t/40-script_openqa-clone-custom-git-refspec.t line 94.
# STDOUT & STDERR:
# openqa-clone-job --skip-chained-deps --within-instance https://openqa.opensuse.org t1234 _GROUP=0 TEST=my_test@user/repo#my_branch BUILD=user/repo#9128 CASEDIR=https://github.com/user/repo.git#my_branch PRODUCTDIR=repo/product NEEDLES_DIR=/my/case/dir/product/needles 
# 
# don't match:
# (?^:openqa-clone-job --skip-chained-deps --within-instance https://openqa.opensuse.org 1234 _GROUP=0 TEST=my_test@user/repo#my_branch BUILD=user/repo#9128 CASEDIR=https://github.com/user/repo.git#my_branch PRODUCTDIR=repo/product NEEDLES_DIR=/my/case/dir/product/needles)
# as expected
ok 30 - command exits successfully # TODO not implemented
# Calling ' unshare -r -n script/openqa-clone-custom-git-refspec https://github.com/user/repo/pull/9128 https://openqa.opensuse.org/t1234,https://openqa.suse.de/t1234'
not ok 31 - multiple short URLs from different hosts point to individual hosts # TODO not implemented
#   Failed (TODO) test 'multiple short URLs from different hosts point to individual hosts'
#   at t/40-script_openqa-clone-custom-git-refspec.t line 96.
# STDOUT & STDERR:
# openqa-clone-job --skip-chained-deps --within-instance https://openqa.opensuse.org t1234 _GROUP=0 TEST=my_test@user/repo#my_branch BUILD=user/repo#9128 CASEDIR=https://github.com/user/repo.git#my_branch PRODUCTDIR=repo/product NEEDLES_DIR=/my/case/dir/product/needles 
# openqa-clone-job --skip-chained-deps --within-instance https://openqa.suse.de t1234 _GROUP=0 TEST=my_test@user/repo#my_branch BUILD=user/repo#9128 CASEDIR=https://github.com/user/repo.git#my_branch PRODUCTDIR=repo/product NEEDLES_DIR=/my/case/dir/product/needles 
# 
# don't match:
# (?^s:openqa-clone-job --skip-chained-deps --within-instance https://openqa.opensuse.org 1234 _GROUP=0 TEST=my_test@user/repo#my_branch BUILD=user/repo#9128 CASEDIR=https://github.com/user/repo.git#my_branch PRODUCTDIR=repo/product NEEDLES_DIR=/my/case/dir/product/needles.* 1234)
# as expected
ok 32 - command exits successfully # TODO not implemented
# Calling ' unshare -r -n script/openqa-clone-custom-git-refspec https://github.com/user/repo/pull/9128'
ok 33 - clone-job command with test from PR description
ok 34 - command exits successfully
# Calling ' unshare -r -n script/openqa-clone-custom-git-refspec https://github.com/user/repo/pull/9128'
ok 35 - clone-job command with test from PR description in list form
ok 36 - command exits successfully
# Calling ' unshare -r -n script/openqa-clone-custom-git-refspec https://github.com/user/repo/pull/9539 https://openqa.opensuse.org/tests/1169326'
ok 37 - clone-job command with multiple URLs in PR and job URL
ok 38 - command exits successfully
ok 39 - no (unexpected) warnings (via done_testing)
1..39
ok
All tests successful.

Test Summary Report
-------------------
t/40-script_openqa-clone-custom-git-refspec.t (Wstat: 0 Tests: 39 Failed: 0)
  TODO passed:   30, 32
Files=1, Tests=39,  5 wallclock secs ( 0.03 usr  0.00 sys +  4.23 cusr  0.32 csys =  4.58 CPU)
Result: PASS
```

Reply:
Ad "PostgreSQL tests": Just below there is a text (see http://open.qa/docs/#testing):
> Or use prove after pointing to a local test database in the environment variable TEST_PG. 
> Also, If you set a custom base directory, be sure to unset it when running tests. Example:
`TEST_PG='DBI:Pg:dbname=openqa_test;host=/dev/shm/tpg' OPENQA_BASEDIR= prove -v t/14-grutasks.t`
I added this commit just to easy writing tests for openQA (maybe it's just myself who spent 80% of time on PR on writing tests, fighting with Codecov and perltidy).